binaries/daemon/src/pending.rs

```diff
@@ -0,0 +1,104 @@
+    /// List of nodes that finished before connecting to the dora daemon.
+    ///
+    /// If this list is non-empty, we should not start the dataflow at all. Instead,
+    /// we report an error to the other nodes.
+    exited_before_subscribe: HashSet<NodeId>,
+}
+
+impl PendingNodes {
+    pub fn insert(&mut self, node_id: NodeId) {
+        self.local_nodes.insert(node_id);
+    }
+
+    pub fn set_external_nodes(&mut self, value: bool) {
+        self.external_nodes = value;
+    }
+
+    pub async fn handle_node_subscription(
+        &mut self,
+        node_id: NodeId,
+        reply_sender: oneshot::Sender<DaemonReply>,
+    ) -> eyre::Result<DataflowStatus> {
+        self.waiting_subscribers
+            .insert(node_id.clone(), reply_sender);
+
+        self.local_nodes.remove(&node_id);
+        if self.local_nodes.is_empty() {
+            if self.external_nodes {
+                Ok(DataflowStatus::LocalNodesReady { success: true })
+            } else {
+                self.answer_subscribe_requests(None).await;
+                Ok(DataflowStatus::AllNodesReady)
+            }
+        } else {
+            Ok(DataflowStatus::LocalNodesPending)
+        }
+    }
+
+    pub async fn handle_node_stop(&mut self, node_id: &NodeId) -> DataflowStatus {
+        self.exited_before_subscribe.insert(node_id.clone());
+
+        self.local_nodes.remove(node_id);
+        if self.local_nodes.is_empty() {
+            if self.external_nodes {
+                DataflowStatus::LocalNodesReady { success: false }
+            } else {
+                self.answer_subscribe_requests(None).await;
+                DataflowStatus::AllNodesReady
+            }
+        } else {
+            // continue waiting for other nodes
+            DataflowStatus::LocalNodesPending
+        }
+    }
+
+    pub async fn handle_external_all_nodes_ready(&mut self, success: bool) {
+        let external_error = if success {
+            None
+        } else {
+            Some("some nodes failed to initalize on remote machines".to_string())
+        };
+        self.answer_subscribe_requests(external_error).await
+    }
+
+    async fn answer_subscribe_requests(&mut self, external_error: Option<String>) {
+        let result = if self.exited_before_subscribe.is_empty() {
+            match external_error {
+                Some(err) => Err(err),
+                None => Ok(()),
+            }
+        } else {
+            Err(format!(
+                "Nodes failed before subscribing: {:?}",
+                self.exited_before_subscribe
+            ))
+        };
+        // answer all subscribe requests
+        let subscribe_replies = std::mem::take(&mut self.waiting_subscribers);
+        for reply_sender in subscribe_replies.into_values() {
+            let _ = reply_sender.send(DaemonReply::Result(result.clone()));
+        }
+    }
+}
+
+pub enum DataflowStatus {
+    AllNodesReady,
+    LocalNodesPending,
+    LocalNodesReady { success: bool },
```

Comment:
Genuinely would prefer to have a `DataflowStatus::Failed` so that we can link all failure to this state. Having a `LocalNodesReady { success: false }` seems unexpected.

Reply:
I pushed some more commits to send the coordinator message inside the `pending` submodule. This way, we can avoid the confusing `LocalNodesReady` state. I also fixed some remaining bugs that I noticed.

Perhaps you could do another round of testing with some Python code to ensure that I didn't break anything by accident?

Reply:
Will do! Thanks a lot!